source/common/buffer/buffer_impl.cc

```diff
@@ -44,171 +62,574 @@ void OwnedImpl::add(const Instance& data) {
+        break;
+      }
+    }
 
-  evbuffer_ptr start_ptr;
-  int rc = evbuffer_ptr_set(buffer_.get(), &start_ptr, start, EVBUFFER_PTR_SET);
-  ASSERT(rc != -1);
+    ASSERT(num_slices_committed > 0);
+  }
+}
 
-  ev_ssize_t copied = evbuffer_copyout_from(buffer_.get(), &start_ptr, data, size);
-  ASSERT(static_cast<uint64_t>(copied) == size);
+void OwnedImpl::copyOut(size_t start, uint64_t size, void* data) const {
+  if (old_impl_) {
+    ASSERT(start + size <= length());
+
+    evbuffer_ptr start_ptr;
+    int rc = evbuffer_ptr_set(buffer_.get(), &start_ptr, start, EVBUFFER_PTR_SET);
+    ASSERT(rc != -1);
+
+    ev_ssize_t copied = evbuffer_copyout_from(buffer_.get(), &start_ptr, data, size);
+    ASSERT(static_cast<uint64_t>(copied) == size);
+  } else {
+    uint64_t bytes_to_skip = start;
+    uint8_t* dest = static_cast<uint8_t*>(data);
+    for (const auto& slice : slices_) {
+      if (size == 0) {
+        break;
+      }
+      uint64_t data_size = slice->dataSize();
+      if (data_size <= bytes_to_skip) {
+        // The offset where the caller wants to start copying is after the end of this slice,
+        // so just skip over this slice completely.
+        bytes_to_skip -= data_size;
+        continue;
+      }
+      uint64_t copy_size = std::min(size, data_size - bytes_to_skip);
+      memcpy(dest, slice->data() + bytes_to_skip, copy_size);
+      size -= copy_size;
+      dest += copy_size;
+      // Now that we've started copying, there are no bytes left to skip over. If there
+      // is any more data to be copied, the next iteration can start copying from the very
+      // beginning of the next slice.
+      bytes_to_skip = 0;
+    }
+    ASSERT(size == 0);
+  }
 }
 
 void OwnedImpl::drain(uint64_t size) {
-  ASSERT(size <= length());
-  int rc = evbuffer_drain(buffer_.get(), size);
-  ASSERT(rc == 0);
+  if (old_impl_) {
+    ASSERT(size <= length());
+    int rc = evbuffer_drain(buffer_.get(), size);
+    ASSERT(rc == 0);
+  } else {
+    while (size != 0) {
+      if (slices_.empty()) {
+        break;
+      }
+      uint64_t slice_size = slices_.front()->dataSize();
+      if (slice_size <= size) {
+        slices_.pop_front();
+        length_ -= slice_size;
+        size -= slice_size;
+      } else {
+        slices_.front()->drain(size);
+        length_ -= size;
+        size = 0;
+      }
+    }
+  }
 }
 
 uint64_t OwnedImpl::getRawSlices(RawSlice* out, uint64_t out_size) const {
-  return evbuffer_peek(buffer_.get(), -1, nullptr, reinterpret_cast<evbuffer_iovec*>(out),
-                       out_size);
+  if (old_impl_) {
+    return evbuffer_peek(buffer_.get(), -1, nullptr, reinterpret_cast<evbuffer_iovec*>(out),
+                         out_size);
+  } else {
+    uint64_t i = 0;
+    for (const auto& slice : slices_) {
+      if (slice->dataSize() == 0) {
+        continue;
+      }
+      if (i < out_size) {
+        out[i].mem_ = slice->data();
+        out[i].len_ = slice->dataSize();
+      }
```

Comment:
We have to loop through the all the slices because of the pre-existing semantics defined for the method in [include/envoy/buffer/buffer.h](https://github.com/envoyproxy/envoy/blob/master/include/envoy/buffer/buffer.h#L131)
```
   * @return the actual number of slices needed, which may be greater than out_size. Passing
   *         nullptr for out and 0 for out_size will just return the size of the array needed
   *         to capture all of the slice data.
```
I'll add a comment in the impl to explain why it does the counterintuitive thing.